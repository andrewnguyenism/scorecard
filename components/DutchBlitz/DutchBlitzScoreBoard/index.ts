export { DutchBlitzScoreBoard } from "./DutchBlitzScoreBoard";
export { DutchBlitzScoreBoardHeader } from "./DutchBlitzScoreBoardHeader";
export { DutchBlitzPlayer, DutchBlitzScoreBoardPlayer } from "./DutchBlitzScoreBoardPlayer";