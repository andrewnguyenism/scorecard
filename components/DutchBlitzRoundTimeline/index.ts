export { DutchBlitzRoundTimeline } from './DutchBlitzRoundTimeline'

export type { DutchBlitzRound } from './DutchBlitzRoundTimeline'
