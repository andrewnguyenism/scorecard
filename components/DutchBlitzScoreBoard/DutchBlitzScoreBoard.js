import { createRef, forwardRef } from "react";
import { GameIdWidget } from "../GameIdWidget";
import { LiveUpdatingRows } from "../LiveUpdatingRows";

export const DutchBlitzScoreBoardHeader = ({ round }) => (
  <div className="flex items-center px-4 py-2 text-gray-600 bg-gray-200 text-sm">
    <div className="w-8/12 text-gray-800 font-bold">Round {round}</div>
    <div className="w-12 text-center">Dutch</div>
    <div className="w-12 text-center">Blitz</div>
    <div className="w-12 text-center">Total</div>
  </div>
)

export const DutchBlitzScoreBoardPlayer = forwardRef(({ currentUserId, place, player }, ref) => (
  <div
    className="flex items-center px-4 py-2"    
    ref={ref}
  >
    <div className="flex items-center w-8/12">
      {place && place >= 0 && place <= 2 && (
        <span className="w-8 font-bold">
          {place == 1 && <span>ðŸ¥‡</span>}
          {place == 2 && <span>ðŸ¥ˆ</span>}
          {place == 3 && <span>ðŸ¥‰</span>}
        </span>
      )}
      {player.name}
      {currentUserId && player.id === currentUserId && <span className="text-xs ml-1">(You)</span>}
    </div>
    <div className="w-12 text-center">{player.dutchScore || 0}</div>
    <div className="w-12 text-center">{player.blitzScore || 0}</div>
    <div className="w-12 text-center text-lg font-bold">
      {player.totalScore || 0}
    </div>
  </div>
));

export const DutchBlitzScoreBoard = ({
  blitz,
  canBlitz,
  canFinishGame,
  canNextRound,
  currentUserId,
  finished,
  finishGame,
  gameId,
  isAdmin,
  nextRound,
  players,
  round,
}) => (
  <div className="m-4 rounded-2xl overflow-hidden border-2">
    <div className="border-b border-gray-300 px-4 py-3 text-gray-600 bg-gray-200 text-lg text-center">
      <div className="text-sm font-bold uppercase mb-2">Dutch Blitz</div>
      <GameIdWidget gameId={gameId} />
      {finished && <div className="text-xs font-bold uppercase mt-2">Finished</div>}
      {!finished && <div className="text-xs text-red-800 font-bold uppercase mt-2">Live</div>}
    </div>
    <DutchBlitzScoreBoardHeader round={round} />
    <LiveUpdatingRows>
      {players.map((player, index,) => (
        <DutchBlitzScoreBoardPlayer currentUserId={currentUserId} key={player.id} player={player} place={index + 1} ref={createRef()} />
      ))}
    </LiveUpdatingRows>
    {canBlitz && (
      <div className="border-t border-gray-300 px-4 py-3 text-gray-600">
        <button
          className="block bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-2xl font-semibold mx-auto px-4 py-2 rounded-2xl"
          onClick={() => blitz()}
        >
          Blitz!
        </button>
      </div>
    )}
    {isAdmin && (canNextRound || canFinishGame) && (
      <div className="flex justify-center border-t border-gray-300 px-4 py-3 text-gray-600 bg-gray-200">
        {canNextRound && (
          <button
            className="bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-sm font-semibold mx-auto px-4 py-2 rounded-2xl"
            onClick={() => nextRound()}
          >
            Next Round
          </button>
        )}
        {canFinishGame && (
          <button
            className="bg-red-700 hover:bg-red-600 text-white uppercase text-sm mx-auto px-4 py-2 rounded-2xl"
            onClick={() => finishGame()}
          >
            Finish Game
          </button>
        )}
      </div>
    )}
  </div>
);
