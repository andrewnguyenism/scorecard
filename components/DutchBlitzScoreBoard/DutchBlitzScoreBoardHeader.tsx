import { FunctionComponent } from "react";

interface Props {
  round: number;
}

export const DutchBlitzScoreBoardHeader: FunctionComponent<Props> = ({
  round,
}) => (
  <div className="flex items-center px-4 py-2 text-gray-600 bg-gray-200 text-sm">
    <div className="w-8/12 text-gray-800 font-bold">Round {round}</div>
    <div className="w-12 text-center">Dutch</div>
    <div className="w-12 text-center">Blitz</div>
    <div className="w-12 text-center">Total</div>
  </div>
);
