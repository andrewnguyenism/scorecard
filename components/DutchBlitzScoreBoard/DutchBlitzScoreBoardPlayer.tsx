import { forwardRef } from 'react'

interface Props {
  blitzScore: number
  dutchScore: number
  name: string
  place?: number
  playerIsCurrentUser?: boolean
  totalScore: number
}

export const DutchBlitzScoreBoardPlayer = forwardRef<HTMLDivElement, Props>(
  (
    { blitzScore, dutchScore, name, place, playerIsCurrentUser, totalScore },
    ref
  ) => (
    <div className="flex items-center px-4 py-2" ref={ref}>
      <div className="flex items-center w-8/12">
        {place && place >= 0 && place <= 2 && (
          <span className="w-8 font-bold">
            {place === 1 && <span>ðŸ¥‡</span>}
            {place === 2 && <span>ðŸ¥ˆ</span>}
            {place === 3 && <span>ðŸ¥‰</span>}
          </span>
        )}
        {name}
        {playerIsCurrentUser && <span className="text-xs ml-1">(You)</span>}
      </div>
      <div className="w-12 text-center">{dutchScore || 0}</div>
      <div className="w-12 text-center">{blitzScore || 0}</div>
      <div className="w-12 text-center text-lg font-bold">
        {totalScore || 0}
      </div>
    </div>
  )
)

DutchBlitzScoreBoardPlayer.displayName = 'DutchBlitzScoreBoardPlayer'
