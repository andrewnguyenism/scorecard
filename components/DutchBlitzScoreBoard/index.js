export {
  DutchBlitzScoreBoardHeader,
  DutchBlitzScoreBoardPlayer,
  DutchBlitzScoreBoard,
} from "./DutchBlitzScoreBoard";
