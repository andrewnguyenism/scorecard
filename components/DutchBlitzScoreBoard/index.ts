export { DutchBlitzScoreBoard } from './DutchBlitzScoreBoard'
export { DutchBlitzScoreBoardHeader } from './DutchBlitzScoreBoardHeader'
export { DutchBlitzScoreBoardPlayer } from './DutchBlitzScoreBoardPlayer'
