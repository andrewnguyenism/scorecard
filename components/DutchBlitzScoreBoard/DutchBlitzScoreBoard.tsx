import { createRef, FunctionComponent } from 'react'

import { GameIdWidget } from '../GameIdWidget'
import { LiveUpdatingRows } from '../LiveUpdatingRows'
import { DutchBlitzScoreBoardHeader } from './DutchBlitzScoreBoardHeader'
import {
  DutchBlitzPlayer,
  DutchBlitzScoreBoardPlayer,
} from './DutchBlitzScoreBoardPlayer'

interface Props {
  blitz: () => void
  canBlitz: boolean
  canFinishGame: boolean
  canNextRound: boolean
  currentUserId: string
  finished: boolean
  finishGame: () => void
  gameId: string
  isAdmin: boolean
  nextRound: () => void
  players: DutchBlitzPlayer[]
  round: number
}

export const DutchBlitzScoreBoard: FunctionComponent<Props> = ({
  blitz,
  canBlitz,
  canFinishGame,
  canNextRound,
  currentUserId,
  finished,
  finishGame,
  gameId,
  isAdmin,
  nextRound,
  players,
  round,
}) => (
  <div className="m-4 rounded-2xl overflow-hidden border-2">
    <div className="border-b border-gray-300 px-4 py-3 text-gray-600 bg-gray-200 text-lg text-center">
      <div className="text-sm font-bold uppercase mb-2">Dutch Blitz</div>
      <GameIdWidget gameId={gameId} />
      {finished && (
        <div className="text-xs font-bold uppercase mt-2">Finished</div>
      )}
      {!finished && (
        <div className="text-xs text-red-800 font-bold uppercase mt-2">
          Live
        </div>
      )}
    </div>
    <DutchBlitzScoreBoardHeader round={round} />
    <LiveUpdatingRows>
      {players.map((player, index) => (
        <DutchBlitzScoreBoardPlayer
          currentUserId={currentUserId}
          key={player.id}
          player={player}
          place={index + 1}
          ref={createRef()}
        />
      ))}
    </LiveUpdatingRows>
    {canBlitz && (
      <div className="border-t border-gray-300 px-4 py-3 text-gray-600">
        <button
          className="block bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-2xl font-semibold mx-auto px-4 py-2 rounded-2xl"
          onClick={() => blitz()}
        >
          Blitz!
        </button>
      </div>
    )}
    {isAdmin && (canNextRound || canFinishGame) && (
      <div className="flex justify-center border-t border-gray-300 px-4 py-3 text-gray-600 bg-gray-200">
        {canNextRound && (
          <button
            className="bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-sm font-semibold mx-auto px-4 py-2 rounded-2xl"
            onClick={() => nextRound()}
          >
            Next Round
          </button>
        )}
        {canFinishGame && (
          <button
            className="bg-red-700 hover:bg-red-600 text-white uppercase text-sm mx-auto px-4 py-2 rounded-2xl"
            onClick={() => finishGame()}
          >
            Finish Game
          </button>
        )}
      </div>
    )}
  </div>
)
