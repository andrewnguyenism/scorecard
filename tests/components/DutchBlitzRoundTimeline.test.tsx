import user from '@testing-library/user-event'
import { render, screen } from '../test-utils'

import { DutchBlitzScoreBoard } from '@/components/DutchBlitzScoreBoard'

describe('DutchBlitzScoreBoard component', () => {
  it('renders', () => {
    const { asFragment } = render(
      <DutchBlitzScoreBoard
        blitz={jest.fn()}
        round={7}
        canBlitz
        canFinishGame
        canNextRound={false}
        currentUserId="test-user-id"
        finishGame={jest.fn()}
        finished={false}
        gameId="AB123"
        isAdmin
        nextRound={jest.fn()}
        players={[{
          blitzScore: 0,
          createdAt: 1609481440,
          dutchScore: 10,
          id: 'player-1',
          name: 'Player 1',
          totalScore: 10,
          updatedAt: 1609481440,
        }, {
          blitzScore: 5,
          createdAt: 1609481440,
          dutchScore: 15,
          id: 'player-2',
          name: 'Player 2',
          totalScore: 5,
          updatedAt: 1609481440,
        }]}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('calls blitz when "Blitz" is clicked', () => {
    const blitz = jest.fn()
    render(
      <DutchBlitzScoreBoard
        blitz={blitz}
        round={7}
        canBlitz
        canFinishGame
        canNextRound
        currentUserId="test-user-id"
        finishGame={jest.fn()}
        finished={false}
        gameId="AB123"
        isAdmin
        nextRound={jest.fn()}
        players={[{
          blitzScore: 0,
          createdAt: 1609481440,
          dutchScore: 10,
          id: 'player-1',
          name: 'Player 1',
          totalScore: 10,
          updatedAt: 1609481440,
        }, {
          blitzScore: 5,
          createdAt: 1609481440,
          dutchScore: 15,
          id: 'player-2',
          name: 'Player 2',
          totalScore: 5,
          updatedAt: 1609481440,
        }]}
      />
    )

    user.click(screen.getByText(/blitz!/i))

    expect(blitz).toBeCalled()
  })

  it('calls finishGame when "Finish Game" is clicked', () => {
    const finishGame = jest.fn()
    render(
      <DutchBlitzScoreBoard
        blitz={jest.fn()}
        round={7}
        canBlitz
        canFinishGame
        canNextRound
        currentUserId="test-user-id"
        finishGame={finishGame}
        finished={false}
        gameId="AB123"
        isAdmin
        nextRound={jest.fn()}
        players={[{
          blitzScore: 0,
          createdAt: 1609481440,
          dutchScore: 10,
          id: 'player-1',
          name: 'Player 1',
          totalScore: 10,
          updatedAt: 1609481440,
        }, {
          blitzScore: 5,
          createdAt: 1609481440,
          dutchScore: 15,
          id: 'player-2',
          name: 'Player 2',
          totalScore: 5,
          updatedAt: 1609481440,
        }]}
      />
    )

    user.click(screen.getByText(/finish game/i))

    expect(finishGame).toBeCalled()
  })

  it('calls nextRound when "Next Round" is clicked', () => {
    const nextRound = jest.fn()
    render(
      <DutchBlitzScoreBoard
        blitz={jest.fn()}
        round={7}
        canBlitz
        canFinishGame
        canNextRound
        currentUserId="test-user-id"
        finishGame={jest.fn()}
        finished={false}
        gameId="AB123"
        isAdmin
        nextRound={nextRound}
        players={[{
          blitzScore: 0,
          createdAt: 1609481440,
          dutchScore: 10,
          id: 'player-1',
          name: 'Player 1',
          totalScore: 10,
          updatedAt: 1609481440,
        }, {
          blitzScore: 5,
          createdAt: 1609481440,
          dutchScore: 15,
          id: 'player-2',
          name: 'Player 2',
          totalScore: 5,
          updatedAt: 1609481440,
        }]}
      />
    )

    user.click(screen.getByText(/next round/i))

    expect(nextRound).toBeCalled()
  })
})
