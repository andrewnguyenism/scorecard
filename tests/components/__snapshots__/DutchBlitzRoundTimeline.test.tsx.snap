// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DutchBlitzScoreBoard component renders 1`] = `
<DocumentFragment>
  <div
    class="m-4 rounded-2xl overflow-hidden border-2"
  >
    <div
      class="border-b border-gray-300 px-4 py-3 text-gray-600 bg-gray-200 text-lg text-center"
    >
      <div
        class="text-sm font-bold uppercase mb-2"
      >
        Dutch Blitz
      </div>
      <div
        class="cursor-pointer rounded-2xl bg-gray-300 hover:bg-gray-400 px-3 inline-flex items-center font-bold"
        role="button"
      >
        <span
          class="text-gray-800 text-xl font-mono font-bold tracking-widest"
        >
          AB123
        </span>
        <svg
          class="h-4 w-4 ml-0.5"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 2a1 1 0 000 2h2a1 1 0 100-2H8z"
          />
          <path
            d="M3 5a2 2 0 012-2 3 3 0 003 3h2a3 3 0 003-3 2 2 0 012 2v6h-4.586l1.293-1.293a1 1 0 00-1.414-1.414l-3 3a1 1 0 000 1.414l3 3a1 1 0 001.414-1.414L10.414 13H15v3a2 2 0 01-2 2H5a2 2 0 01-2-2V5zM15 11h2a1 1 0 110 2h-2v-2z"
          />
        </svg>
      </div>
      <div
        class="animate-pulse text-xs text-red-800 font-bold uppercase mt-2"
      >
        Live
      </div>
    </div>
    <div
      class="flex items-center px-4 py-2 text-gray-600 bg-gray-200 text-sm"
    >
      <div
        class="w-8/12 text-gray-800 font-bold"
      >
        Round 7
      </div>
      <div
        class="w-12 text-center"
      >
        Dutch
      </div>
      <div
        class="w-12 text-center"
      >
        Blitz
      </div>
      <div
        class="w-12 text-center"
      >
        Total
      </div>
    </div>
    <div>
      <div
        class="flex items-center px-4 py-2"
      >
        <div
          class="flex items-center w-8/12"
        >
          <span
            class="w-8 font-bold"
          >
            <span>
              ðŸ¥‡
            </span>
          </span>
          Player 1
        </div>
        <div
          class="w-12 text-center"
        >
          10
        </div>
        <div
          class="w-12 text-center"
        >
          0
        </div>
        <div
          class="w-12 text-center text-lg font-bold"
        >
          10
        </div>
      </div>
      <div
        class="flex items-center px-4 py-2"
      >
        <div
          class="flex items-center w-8/12"
        >
          <span
            class="w-8 font-bold"
          >
            <span>
              ðŸ¥ˆ
            </span>
          </span>
          Player 2
        </div>
        <div
          class="w-12 text-center"
        >
          15
        </div>
        <div
          class="w-12 text-center"
        >
          5
        </div>
        <div
          class="w-12 text-center text-lg font-bold"
        >
          5
        </div>
      </div>
    </div>
    <div
      class="border-t border-gray-300 px-4 py-3 text-gray-600"
    >
      <button
        class="block bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-2xl font-semibold mx-auto px-4 py-2 rounded-2xl"
      >
        Blitz!
      </button>
    </div>
    <div
      class="flex justify-center border-t border-gray-300 px-4 py-3 text-gray-600 bg-gray-200"
    >
      <button
        class="bg-red-700 hover:bg-red-600 text-white uppercase text-sm mx-auto px-4 py-2 rounded-2xl"
      >
        Finish Game
      </button>
    </div>
  </div>
</DocumentFragment>
`;
